host/windows/syscall.c

```diff
@@ -480,142 +1103,519 @@ ssize_t oe_syscall_writev_ocall(
+
+    // Find file next doesn't return '.' because it shows up in opendir and we
+    // lose it but we know it is there, so we can just return it
+    if (pdir->dir_offs == 0)
+    {
+        entry->d_off = pdir->dir_offs++;
+        entry->d_type = OE_DT_DIR;
+        entry->d_reclen = sizeof(struct oe_dirent);
+        entry->d_name[0] = '.';
+        entry->d_name[1] = '\0';
+        ret = 0;
+        goto done;
+    }
+
+    if (!FindNextFileW(pdir->hFind, &pdir->FindFileData))
+    {
+        DWORD winerr = GetLastError();
+
+        if (winerr == ERROR_NO_MORE_FILES)
+        {
+            /* Return 1 to indicate there no more entries. */
+            ret = 1;
+        }
+        else
+        {
+            _set_errno(_winerr_to_errno(winerr));
+            ret = -1;
+        }
+        goto done;
+    }
+
+    nlen = WideCharToMultiByte(
+        CP_UTF8, 0, pdir->FindFileData.cFileName, -1, NULL, 0, NULL, NULL);
+    (void)WideCharToMultiByte(
+        CP_UTF8,
+        0,
+        pdir->FindFileData.cFileName,
+        nlen,
+        entry->d_name,
+        sizeof(entry->d_name),
+        NULL,
+        NULL);
+
+    if(nlen == 0)
+    {
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+
+    entry->d_type = 0;
+    if (pdir->FindFileData.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+    {
+        entry->d_type = OE_DT_DIR;
+    }
+    else if (pdir->FindFileData.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)
+    {
+        entry->d_type = OE_DT_LNK;
+    }
+    else
+    {
+        entry->d_type = OE_DT_REG;
+    }
+
+    entry->d_off = pdir->dir_offs++;
+    entry->d_reclen = sizeof(struct oe_dirent);
+
+    ret = 0;
+
+done:
+    return ret;
 }
 
 void oe_syscall_rewinddir_ocall(uint64_t dirp)
 {
-    OE_UNUSED(dirp);
+    DWORD err = 0;
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
+    WCHAR* wpathname = pdir->pdirpath;
+    // Undo abosolute path forcing again. We do this over because we need to
+    // preserve the allocation address for free.
+    if (wcslen(wpathname) >= 3 && wpathname[0] == '/' && wpathname[2] == ':')
+    {
+        wpathname++;
+    }
 
-    PANIC;
+    if(!FindClose(pdir->hFind))
+    {
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+
+    memset(&pdir->FindFileData, 0, (size_t)sizeof(pdir->FindFileData));
+
+    pdir->hFind = FindFirstFileW(wpathname, &pdir->FindFileData);
+    if (pdir->hFind == INVALID_HANDLE_VALUE)
+    {
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+    pdir->dir_offs = 0;
+
+done:
+    return;
 }
 
 int oe_syscall_closedir_ocall(uint64_t dirp)
 {
-    OE_UNUSED(dirp);
+    int ret = -1;
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
 
-    PANIC;
+    if (!dirp)
+    {
+        goto done;
+    }
+    if (!FindClose(pdir->hFind))
+    {
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+
+    free(pdir->pdirpath);
```

Comment:
Is it guaranteed that pdir->pdirpath is not null?

Reply:
Now free() can deal with NULL.